sirepo/pkcli/raydata.py

```diff
@@ -18,5 +156,455 @@ def run_background(cfg_dir):
+    # beamline. The callback given to subscribe would then call into this daemon letting us know that
+    # new documents are available.
+    # But, for now it is easiest to just poll
+    async def _poll_catalog_for_scans(catalog_name):
+        def _collect_new_scans_and_queue(last_known):
+            l = last_known
+            m = _Metadata(last_known)
+            r = [
+                scan
+                for uid, scan in databroker.catalog[catalog_name]
+                .search(databroker.queries.TimeRange(since=m.start()))
+                .items()
+                if uid != m.uid()
+            ]
+            r.sort(key=lambda x: _Metadata(x).start())
+            for s in r:
+                if _Metadata(s).is_scan_plan_executing():
+                    return l
+                _queue_for_analysis(s, catalog_name)
+                l = s
+            return l
+
+        async def _poll_for_new_scans(most_recent):
+            while True:
+                most_recent = _collect_new_scans_and_queue(most_recent)
+                await pkasyncio.sleep(2)
+
+        pkdlog("catalog_name={}", catalog_name)
+        c = None
+        while not c:
+            try:
+                c = databroker.catalog[catalog_name]
+            except KeyError:
+                pkdlog(f"no catalog_name={catalog_name}. Retrying...")
+                await pkasyncio.sleep(15)
+        s = c[-1]
+        if not _Analysis.have_analyzed_scan(_Metadata(s).uid(), catalog_name):
+            _queue_for_analysis(s, catalog_name)
+        await _poll_for_new_scans(s)
+        raise AssertionError("should never get here")
+
+    def _queue_for_analysis(scan, catalog_name):
+        s = PKDict(
+            uid=scan if isinstance(scan, str) else _Metadata(scan).uid(),
+            catalog_name=catalog_name,
+        )
+        pkdlog("scan={}", s)
+        if s not in _SCANS_AWAITING_ANALYSIS:
+            pkio.unchecked_remove(_Analysis.analysis_output_dir(s.uid).join("*"))
+            _SCANS_AWAITING_ANALYSIS.append(s)
+            _Analysis.set_scan_status(s, _AnalysisStatus.PENDING)
+
+    def _scan_info(catalog, scan_uuid, scans_data, status=None):
+        m = _Metadata(catalog[scan_uuid])
+        d = PKDict(
+            uid=scan_uuid,
+            status=status,
+            pdf=True if _analysis_pdf_paths(scan_uuid) else False,
+        )
+        for c in _DEFAULT_COLUMNS:
+            d[c] = getattr(m, c)()
+
+        for c in scans_data.get("selectedColumns", []):
+            d[c] = m.get_scan_field(c)
+        return d
+
+    def _scan_info_result(scans):
+        return PKDict(
+            data=PKDict(
+                scans=sorted(scans, key=lambda e: e.start),
+                cols=[k for k in scans[0].keys() if k not in _NON_DISPLAY_SCAN_FIELDS]
+                if scans
+                else [],
+            )
+        )
+
+    sirepo.srtime.init_module()
+    pkio.mkdir_parent(sirepo.feature_config.cfg().raydata.db_dir)
```

Comment:
That loop was a "fail fast" check to make sure that we had a notebook for each catalog we are monitoring. If you dig down into the call in pykern you'll see we raise not found if the file doesn't exist.

I think the check is worthwhile. But, it could be put in a clearly named function to make it easier for readers to understand what its use is.

Reply:
e799cdc8671eaf3a2fa69713cd0682b70a4ea7bb